Source/WebCore/dom/Element.cpp

```diff
@@ -2735,232 +2735,307 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
+    }
 }
 
-void Element::updateEffectiveDirectionalityOfDirAuto()
+bool Element::hasValidTextDirectionState() const
 {
-    auto result = computeDirectionalityFromText();
-    setHasDirAutoFlagRecursively(this, true, result.strongDirectionalityNode.get());
-    updateEffectiveDirectionality(result.direction);
-    if (renderer() && renderer()->style().direction() != result.direction)
-        invalidateStyleForSubtree();
+    return elementHasValidTextDirectionState(*this);
 }
 
-auto Element::computeDirectionalityFromText() const -> TextDirectionWithStrongDirectionalityNode
+bool Element::hasAutoTextDirectionState() const
 {
-    if (RefPtr textControl = dynamicDowncast<HTMLTextFormControlElement>(const_cast<Element*>(this))) {
-        if (textControl->dirAutoUsesValue()) {
-            auto direction = textControl->value().defaultWritingDirection();
-            if (!direction)
-                return { TextDirection::LTR, nullptr };
-            return { *direction == U_LEFT_TO_RIGHT ? TextDirection::LTR : TextDirection::RTL, WTFMove(textControl) };
-        }
-    }
+    return elementHasAutoTextDirectionState(*this);
+}
 
-    RefPtr node = firstChild();
-    while (node) {
-        // Skip bdi, script, style and text form controls.
-        auto* element = dynamicDowncast<Element>(*node);
-        if (node->hasTagName(bdiTag) || node->hasTagName(scriptTag) || node->hasTagName(styleTag)
-            || (element && element->isTextField())) {
-            node = NodeTraversal::nextSkippingChildren(*node, this);
+std::optional<TextDirection> Element::computeAutoTextDirection() const
+{
+    if (auto* textFormControl = dynamicDowncast<HTMLTextFormControlElement>(*this))
+        return computeTextFormControlElementTextDirection(*textFormControl);
+
+    if (auto* slotElement = dynamicDowncast<HTMLSlotElement>(*this))
+        return computeSlotElementTextDirection(*slotElement);
+
+    return computeElementContainedTextAutoDirection(*this);
+}
+
+std::optional<TextDirection> Element::computeTextDirection() const
+{
+    if (!(selfOrPrecedingNodesAffectDirAuto() && hasAutoTextDirectionState()))
+        return std::nullopt;
+    return computeElementTextDirection(*this, elementTextDirectionState(*this));
+}
+
+void Element::updateEffectiveTextDirectionOfDescendants(std::optional<TextDirection> direction, Element* exclude)
+{
+    for (auto it = descendantsOfType<Element>(*this).begin(); it; ) {
+        Ref child = *it;
+
+        if (elementHasValidTextDirectionState(child)) {
+            it.traverseNextSkippingChildren();
             continue;
         }
 
-        // Skip elements with valid dir attribute
-        if (element) {
-            if (isValidDirValue(element->attributeWithoutSynchronization(dirAttr))) {
-                node = NodeTraversal::nextSkippingChildren(*element, this);
-                continue;
-            }
-        }
+        if (child.ptr() != exclude)
```

Comment:
This "exclude" business is very confusing. We don't process the excluded element but we process its descendants?

Reply:
> This "exclude" business is very confusing

Since we update the descendants and the ancestors of element _element_, we have to protect _element_ itself from being reset when updating the descendants any of the _element_'s ancestors.

> We don't process the excluded element but we process its descendants?

You are right. This was a bug. I fixed it by calling `traverseNextSkippingChildren()` when we revisit _element_.